New promise typing using unions is less flexible than it should be (and used to be)

For example, this code used to type-check fine (prior to: https://github.com/borisyankov/DefinitelyTyped/commit/c95561be3526f90f0e717734a659d1c01d0492d4 )
```
/// <reference path='es6-promise.d.ts' />

interface T1 {
  __t1: string;
}

interface T2 {
  __t2: string;
}

interface T3 {
  __t3: string;
}

function f1() : Promise<T1> {
  return Promise.resolve({ __t1: "foo_t1" });
}

function f2(x:T1) : T2 {
  return { __t2: x.__t1 + ":foo_21" };
}

var x3 = f1()
  .then(f2, (e:Error) => {
    console.log("error 1");
    throw e;
  })
  .then((x:T2) => {
    return { __t3: x.__t2 + "bar" };
  });
```
But now it gives these errors (in TypeScript 1.4.1): 
```
/Users/ldixon/code/uproxy/uproxy-networking/build/third_party/typings/es6-promise/promise-oddity.ts(23,10): error TS2453: The type argument for type parameter 'U' cannot be inferred from the usage. Consider specifying the type arguments explicitly.
  Type argument candidate 'T2' is not a valid type argument because it is not a supertype of candidate 'void'.
```

--- a/es6-promise/es6-promise-tests.ts
+++ b/es6-promise/es6-promise-tests.ts
@@ -160,3 +160,32 @@ getJSON('story.json').then(function(story: Story) {
 }).then(function() {
   (<HTMLElement>document.querySelector('.spinner')).style.display = 'none';
 });
+
+interface T1 {
+    __t1: string;
+}
+
+interface T2 {
+    __t2: string;
+}
+
+interface T3 {
+    __t3: string;
+}
+
+function f1(): Promise<T1> {
+    return Promise.resolve({ __t1: "foo_t1" });
+}
+
+function f2(x: T1): T2 {
+    return { __t2: x.__t1 + ":foo_21" };
+}
+
+var x3 = f1()
+    .then(f2, (e: Error) => {
+    console.log("error 1");
+    throw e;
+})
+    .then((x: T2) => {
+    return { __t3: x.__t2 + "bar" };
+});

--- a/es6-promise/es6-promise.d.ts
+++ b/es6-promise/es6-promise.d.ts
@@ -4,7 +4,8 @@
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
 
 interface Thenable<R> {
-	then<U>(onFulfilled?: (value: R) => U | Thenable<U>,  onRejected?: (error: any) => U | Thenable<U>): Thenable<U>;
+    then<U>(onFulfilled?: (value: R) => U | Thenable<U>, onRejected?: (error: any) => U | Thenable<U>): Thenable<U>;
+    then<U>(onFulfilled?: (value: R) => U | Thenable<U>, onRejected?: (error: any) => void): Thenable<U>;
 }
 
 declare class Promise<R> implements Thenable<R> {
@@ -27,7 +28,8 @@ declare class Promise<R> implements Thenable<R> {
 	 * @param onFulfilled called when/if "promise" resolves
 	 * @param onRejected called when/if "promise" rejects
 	 */
-	then<U>(onFulfilled?: (value: R) => U | Thenable<U>,  onRejected?: (error: any) => U | Thenable<U>): Promise<U>;
+    then<U>(onFulfilled?: (value: R) => U | Thenable<U>, onRejected?: (error: any) => U | Thenable<U>): Promise<U>;
+    then<U>(onFulfilled?: (value: R) => U | Thenable<U>, onRejected?: (error: any) => void): Promise<U>;
 
 	/**
 	 * Sugar for promise.then(undefined, onRejected)

